amethyst_tiles/src/pass.rs

```
@@ -0,0 +1,468 @@
+            change: Default::default(),
+        }))
+    }
+}
+
+/// `RenderGroup` providing culling, drawing and transparency functionality for 3D `TileMap` components.
+#[derive(Derivative)]
+#[derivative(Debug(bound = ""))]
+pub struct DrawTiles2D<B: Backend, T: Tile, E: CoordinateEncoder> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    env: FlatEnvironmentSub<B>,
+    textures: TextureSub<B>,
+    vertex: DynamicVertexBuffer<B, SpriteArgs>,
+    sprites: OrderedOneLevelBatch<TextureId, SpriteArgs>,
+    change: util::ChangeDetection,
+
+    #[derivative(Debug = "ignore")]
+    _marker: PhantomData<(T, E)>,
+}
+
+impl<B: Backend, T: Tile, E: CoordinateEncoder> RenderGroup<B, World> for DrawTiles2D<B, T, E> {
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        #[cfg(feature = "profiler")]
+        profile_scope!("prepare");
+
+        let mut changed = false;
+        let (sprite_sheet_storage, tex_storage, hiddens, tile_maps) = <(
+            Read<'_, AssetStorage<SpriteSheet>>,
+            Read<'_, AssetStorage<Texture>>,
+            ReadStorage<'_, Hidden>,
+            ReadStorage<'_, TileMap<T, E>>,
+        )>::fetch(world);
+
+        let sprites_ref = &mut self.sprites;
+        let textures_ref = &mut self.textures;
+
+        sprites_ref.swap_clear();
+
+        self.env.process(factory, index, world);
+
+        if let Some((tile_map, _)) = (&tile_maps, !&hiddens).join().next() {
+            // Render the first tilemap we find
+            let (top, mut bottom) = fetch_world_bounds(tile_map, world);
+            bottom.z = 0;
+
+            let camera_region = Region::new(top, bottom);
+            camera_region
+                .iter()
+                .filter_map(|coord| {
+                    let tile = tile_map.get(&coord)?;
```

See my other comment. Should this panic if tile_map.get() return None?

---

I'd prefer not to fail hard for cases like this, and fail softly. Maybe a warning?

---

I've added some debug assertions, but wont be failing on out-of-bounds.